internal/controller/apiextensions/claim/configurator.go

```diff
@@ -183,82 +179,94 @@ func filter(in map[string]any, keys ...string) map[string]any {
 	return out
 }
 
-// APIClaimConfigurator configures the supplied claims with fields
-// from the composite. This includes late-initializing spec values
-// and updating status fields in claim.
-type APIClaimConfigurator struct {
-	client client.Client
-}
+func keep(in map[string]any, keys ...string) map[string]any {
+	filter := map[string]bool{}
+	for _, k := range keys {
+		filter[k] = true
+	}
 
-// NewAPIClaimConfigurator returns a APIClaimConfigurator.
-func NewAPIClaimConfigurator(client client.Client) *APIClaimConfigurator {
-	return &APIClaimConfigurator{client: client}
+	out := map[string]any{}
+	for k, v := range in {
+		if filter[k] {
+			out[k] = v
+		}
+	}
+	return out
 }
 
-// Configure the supplied claims with fields from the composite.
+// configure the supplied claim patch with fields from the composite.
 // This includes late-initializing spec values and updating status fields in claim.
-func (c *APIClaimConfigurator) Configure(ctx context.Context, cm resource.CompositeClaim, cp resource.Composite) error {
-	ucm, ok := cm.(*claim.Unstructured)
-	if !ok {
-		return nil
+func configureClaim(_ context.Context, cm *claim.Unstructured, cmPatch *claim.Unstructured, cp *composite.Unstructured, cpPatch *composite.Unstructured) error { //nolint:gocyclo // Only slightly over (10)
+	existing := cm.GetResourceReference()
+	proposed := meta.ReferenceTo(cpPatch, cpPatch.GetObjectKind().GroupVersionKind())
+	equal := cmp.Equal(existing, proposed, cmpopts.IgnoreFields(corev1.ObjectReference{}, "UID"))
+
+	// We refuse to 're-bind' a claim that is already bound to a different
+	// composite resource.
+	if existing != nil && !equal {
+		return errors.New(errBindClaimConflict)
 	}
-	ucp, ok := cp.(*composite.Unstructured)
-	if !ok {
-		return nil
+
+	cmPatch.SetResourceReference(proposed)
+
+	// If existing claim has the status set,
+	// copy the conditions to the patch
+	if s, ok := cm.Object["status"]; ok {
+		fs, ok := s.(map[string]any)
+		if !ok {
+			return errors.Wrap(errors.New(errUnsupportedSrcObject), errMergeClaimStatus)
+		}
+		cmPatch.Object["status"] = keep(fs, "conditions")
+	} else {
+		cmPatch.Object["status"] = map[string]any{}
```

Comment:
What would be the benefit? `Object["status"]` or `Object["spec"]` are used across configurator.

At this particular line, we need to create an empty status placeholder, so that we can perform merge from the composite later on. Given that we are producing here a patch, `cmPatch.Object` does not contains `status` at all.  Hence, `fieldpath.Pave(cmPatch.Object).GetValueInto("status", map[string]any())` is going to produce an error.